File: src/App.test.js
import {
  Fibonacci,
  FibonacciRecursive,
  FibonacciLoop,
  FibonacciGeneralTerm,
  FibonacciList,
} from "./App.js";

describe("フィボナッチ数列", () => {
  let fib;
  let recursive;
  let loop;
  let generalTerm;
  beforeEach(() => {
    recursive = new Fibonacci(new FibonacciRecursive());
    loop = new Fibonacci(new FibonacciLoop());
    generalTerm = new Fibonacci(new FibonacciGeneralTerm());
  });

  test.each([
    [0, 0],
    [1, 1],
    [2, 1],
    [3, 2],
    [4, 3],
    [5, 5],
  ])("%iを渡したら%pを返す", (number, expected) => {
    expect(recursive.exec(number)).toEqual(expected);
  });

  test("大きな数値を計算する(再起処理による実装)", () => {
    expect(recursive.exec(40)).toEqual(102334155);
  });

  test("大きな数値を計算する(ループ処理による実装)", () => {
    expect(loop.exec(40)).toEqual(102334155);
  });

  test("大きな数値を計算する(一般項による実装)", () => {
    expect(generalTerm.exec(40)).toEqual(102334155);
  });

  test("40までのフィボナッチ配列を返す", () => {
    const command = new Fibonacci(new FibonacciRecursive());
    const listCommand = new FibonacciList(command);
    const result = listCommand.exec(40);

    expect(result[0]).toEqual(1);
    expect(result[result.length - 3]).toEqual(39088169);
    expect(result[result.length - 2]).toEqual(63245986);
    expect(result[result.length - 1]).toEqual(102334155);
  });
});

File: api/app_test.rb
require 'minitest/autorun'
require_relative './app'

class FibonacciTest < Minitest::Test
  def setup
    @recursive = Fibonacci.new(FibonacciRecursive.new)
    @loop = Fibonacci.new(FibonacciLoop.new)
    @general_term = Fibonacci.new(FibonacciGeneralTerm.new)
  end

  def test_fibonacci
    cases = [[0, 0], [1, 1], [2, 1], [3, 2], [4, 3], [5, 5]]
    cases.each do |i|
      assert_equal i[1], @recursive.exec(i[0])
    end
  end

  def test_大きな数字_再起処理による実装
    assert_equal 102_334_155, @recursive.exec(40)
  end

  def test_大きな数字_ループ処理による実装
    assert_equal 102_334_155, @loop.exec(40)
  end

  def test_大きな数字_一般項による実装
    assert_equal 102_334_155, @general_term.exec(40)
  end
end
